testnet/stacks-node/src/nakamoto_node/relayer.rs

```diff
@@ -585,76 +632,22 @@ impl RelayerThread {
-            // This would be true if the stacks tip's tenure is at or descends from this snapshot.
-            // If there has _not_ been a BlockFound, then we should issue one.
-            let ih = self
-                .sortdb
-                .index_handle(&last_winning_snapshot.sortition_id);
-            let need_blockfound = if stacks_tip_sn.block_height > last_winning_snapshot.block_height
-            {
-                // stacks tip is ahead of this snapshot, so no BlockFound can be issued.
-                test_debug!("Relayer: stacks_tip_sn.block_height ({}) > last_winning_snapshot.block_height ({})", stacks_tip_sn.block_height, last_winning_snapshot.block_height);
-                false
-            } else if stacks_tip_sn.block_height == last_winning_snapshot.block_height
-                && stacks_tip_sn.consensus_hash == last_winning_snapshot.consensus_hash
-            {
-                // this is the ongoing tenure snapshot. A BlockFound has already been issued. We
-                // can instead opt to Extend
-                test_debug!(
-                    "Relayer: ongoing tenure {} already represents last-winning snapshot",
-                    &stacks_tip_sn.consensus_hash
-                );
-                self.tenure_extend_timeout = Some(Instant::now());
-                false
-            } else {
-                // stacks tip's snapshot may be an ancestor of the last-won sortition.
-                // If so, then we can issue a BlockFound.
-                SortitionDB::get_ancestor_snapshot(
-                    &ih,
-                    stacks_tip_sn.block_height,
-                    &last_winning_snapshot.sortition_id,
-                )
-                .map_err(|e| {
-                    error!("Relayer: Failed to load ancestor snapshot: {e:?}");
-                    e
-                })
-                .ok()
-                .flatten()
-                .map(|sn| {
-                    let need_blockfound = sn.consensus_hash == stacks_tip_sn.consensus_hash;
-                    if !need_blockfound {
-                        test_debug!(
-                            "Relayer: stacks_tip_sn.consensus_hash ({}) != sn.consensus_hash ({})",
-                            &stacks_tip_sn.consensus_hash,
-                            &sn.consensus_hash
-                        );
-                    }
-                    need_blockfound
-                })
-                .unwrap_or_else(|| {
-                    test_debug!(
-                        "Relayer: no ancestor at height {} off of sortition {} height {}",
-                        stacks_tip_sn.block_height,
-                        &last_winning_snapshot.consensus_hash,
-                        last_winning_snapshot.block_height
-                    );
-                    false
-                })
-            };
-            if need_blockfound {
+            if self.need_block_found(&canonical_stacks_snapshot, &last_winning_snapshot) {
```

Comment:
I like that you separated `need_block_found` into a new function... but it occurs to me that this would also be more clear if we separated the `empty_sortition` and `has_sortition` cases...

```rust
if sn.sortition {
    self.choose_directive_sortition_with_winner(..)
} else {
    self.choose_directive_sortition_without_winner(..)
}
```